packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java

```diff
@@ -10,64 +12,152 @@
 import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
 import io.flutter.plugin.common.MethodChannel.Result;
 import io.flutter.plugin.common.PluginRegistry;
-import java.util.Arrays;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
 /** FirebaseDatabasePlugin */
 public class FirebaseDatabasePlugin implements MethodCallHandler {
 
+  private final MethodChannel channel;
+  private static final String EVENT_TYPE_CHILD_ADDED = "_EventType.childAdded";
+  private static final String EVENT_TYPE_CHILD_REMOVED = "_EventType.childRemoved";
+  private static final String EVENT_TYPE_CHILD_CHANGED = "_EventType.childChanged";
+  private static final String EVENT_TYPE_CHILD_MOVED = "_EventType.childMoved";
+  private static final String EVENT_TYPE_VALUE = "_EventType.value";
+
+  // Handles are ints used as indexes into the sparse array of active observers
+  private int nextHandle = 0;
+  private final SparseArray<EventObserver> observers =
+      new SparseArray<EventObserver>();
+
   public static void registerWith(PluginRegistry.Registrar registrar) {
     final MethodChannel channel =
         new MethodChannel(registrar.messenger(), "plugins.flutter.io/firebase_database");
     channel.setMethodCallHandler(new FirebaseDatabasePlugin(channel));
   }
 
-  private FirebaseDatabasePlugin(final MethodChannel channel) {
-    FirebaseDatabase.getInstance()
-        .getReference()
-        .limitToLast(10)
-        .addChildEventListener(
-            new ChildEventListener() {
-              @Override
-              public void onCancelled(DatabaseError error) {}
-
-              @Override
-              public void onChildAdded(DataSnapshot snapshot, String previousChildName) {
-                List arguments = Arrays.asList(snapshot.getKey(), snapshot.getValue());
-                channel.invokeMethod("DatabaseReference#childAdded", arguments);
-              }
-
-              @Override
-              public void onChildChanged(DataSnapshot snapshot, String previousChildName) {}
-
-              @Override
-              public void onChildMoved(DataSnapshot snapshot, String previousChildName) {}
-
-              @Override
-              public void onChildRemoved(DataSnapshot snapshot) {}
-            });
+  private FirebaseDatabasePlugin(MethodChannel channel) {
+    this.channel = channel;
+  }
+
+  private static DatabaseReference getReference(Map<String, ?> arguments) {
+    @SuppressWarnings("unchecked")
+    String path = (String) arguments.get("path");
+    DatabaseReference reference = FirebaseDatabase.getInstance().getReference();
+    if (path != null)
+      reference = reference.child(path);
```

Comment:
Yeah I think Travis is going to make me dartfmt everything so unless I want to use curly braces I'll have to obey this convention.